Make sure we always de-serialize transparent functions.

eeckstein opened the pull request:
We didn’t do that if an optimization looked up a witness table.

Fixes rdar://problem/30544344

I couldn’t come up with an isolated test case, but this should be covered with our existing tests.
(The problem shows up when inlining of generics are enabled)

eeckstein committed:
Make sure we always de-serialize transparent functions.
We didn’t do that if an optimization looked up a witness table.

Fixes rdar://problem/30544344

I couldn’t come up with an isolated test case, but this should be covered with our existing tests.
(The problem shows up when inlining of generics are enabled)

eeckstein commented:
@swift-ci Please smoke test and merge

jrose-apple commented:
Per recent conversations, is "transparent" the right test? We're using this as a stand-in for "does not have a public symbol", right?

slavapestov commented:
We should revisit this soon. I suspect the right test is 'fragile' and also we should always emit fragile functions with shared linkage, not public.

jrose-apple commented:
This is the big case-by-case @swiftix and I made. We need to turn that into a document somewhere.

eeckstein commented:
@jrose-apple transparent is ok, IMO, because we are not generating code for transparent functions in the module which defines them (That's what I implemented some time ago)

jrose-apple commented:
@eeckstein That's *mostly* true, but transparent functions aren't going to be the only such functions. The new `@inlinable` will do that too under some circumstances.

eeckstein commented:
OK, but then we should change it when inlinable will be implemented (AFAIK it's not done yet).
Beside that, I think this behavior (not emitting code for a function) only makes sense for always-inline functions (like transparent).

slavapestov commented:
Before we can change the linkage of inlineable functions, we need to fix the 'merge modules' step to not throw away SIL. I think that might be the only remaining piece though. Can you think of anything else @eeckstein?

eeckstein commented:
@slavapestov I don't know the details here

slavapestov commented:
@jrose-apple I thought @inlineable was always going to emit into the client (and also into the library if its referenced there, but only with shared linkage)

slavapestov commented:
Ditto for @inline(__always)

eeckstein commented:
Currently @inline(__always) is just an optimization hint. So it should not affect the linkage

slavapestov commented:
@inline(__always) also forces serialization of the function (making it fragile) and enables inlineable diagnostics for the body. So it's like @_inlineable except it *also* is a hint to the performance inliner to inline the function. So I'm not sure what we want here.

slavapestov commented:
I kind of want @inline(__always) to replace @_transparent, except possibly for a few cases in the stdlib...

eeckstein commented:
I didn't know that. This must have been added later

slavapestov commented:
Yeah, I added all this recently when I implemented @_inlineable :)

swiftix commented:
We discussed with @eeckstein even before he submitted the PR that transparent checks are too narrow at the moment and we need a more general check which would cover the cases you mentioned. But the aim of this specific PR is not solve all those issues, but to unblock the inlining of generics. We should cover the transparent checks in separate PRs, most likely in one go over the whole optimizer/compiler, so that we can fix all places where it is required.

eeckstein committed:
function signature opts: Have an even more stricter check if a specialized function with that name already exists.

It’s not enough to just check for public functions in other modules.
In case of fragile functions, it can also be a non-public function in another module which potentially interferes with the specialization we want to create.

Anyway, this is just a workaround until we fix the mangling for function-signature opts. It should always produce unique names.

eeckstein commented:
@swift-ci Please smoke test and merge

eeckstein commented:
@swift-ci Please smoke test and merge

eeckstein commented:
@swift-ci Please smoke test and merge

shahmishal commented:
```
FAILED: benchmark/O-armv7k-apple-watchos2.0/Sim2DArray.o 
cd /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/benchmark && /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/swiftc -c -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/WatchOS.platform/Developer/SDKs/WatchOS3.2.sdk -target armv7k-apple-watchos2.0 -F /Applications/Xcode-beta.app/Contents/Developer/Platforms/WatchOS.platform/Developer/SDKs/WatchOS3.2.sdk/../../../Developer/Library/Frameworks -O -D INTERNAL_CHECKS_ENABLED -no-link-objc-runtime -I /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/benchmark/utils/ObjectiveCTests -parse-as-library -whole-module-optimization -module-name Sim2DArray -emit-module-path /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/benchmark/O-armv7k-apple-watchos2.0/Sim2DArray.swiftmodule -I /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/benchmark/O-armv7k-apple-watchos2.0 -o /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/benchmark/O-armv7k-apple-watchos2.0/Sim2DArray.o /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/benchmark/single-source/Sim2DArray.swift
Assertion failed: (isa<X>(Val) && "cast<Ty>() argument of incompatible type!"), function cast, file /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/llvm/include/llvm/Support/Casting.h, line 229.
0  swift                    0x000000010b8bbb38 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  swift                    0x000000010b8bc1e6 SignalHandler(int) + 454
2  libsystem_platform.dylib 0x00007fffb3eb8bba _sigtramp + 26
3  swift                    0x00000001095779f5 swift::SubstitutionMap::addConformance(swift::CanType, swift::ProtocolConformanceRef) + 181
4  libsystem_c.dylib        0x00007fffb3d3f420 abort + 129
5  libsystem_c.dylib        0x00007fffb3d06893 basename_r + 0
6  swift                    0x0000000108f1c516 (anonymous namespace)::MemoryToRegisters::run() + 10134
7  swift                    0x0000000108f19d34 (anonymous namespace)::SILMem2Reg::run() + 468
8  swift                    0x0000000108e34c92 swift::SILPassManager::runPassOnFunction(swift::SILFunctionTransform*, swift::SILFunction*) + 2258
9  swift                    0x0000000108e35b1a swift::SILPassManager::runFunctionPasses(llvm::ArrayRef<swift::SILFunctionTransform*>) + 1082
10 swift                    0x0000000108e36e34 swift::SILPassManager::runOneIteration() + 948
11 swift                    0x000000010873423b swift::SILPassManager::executePassPipelinePlan(swift::SILPassPipelinePlan const&) + 187
12 swift                    0x0000000108e3f882 swift::runSILOptimizationPasses(swift::SILModule&) + 114
13 swift                    0x00000001085eea7d swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 13037
14 swift                    0x00000001085ab150 main + 3312
15 libdyld.dylib            0x00007fffb3cab255 start + 1
16 libdyld.dylib            0x000000000000001b start + 1278561735
Stack dump:
0.	Program arguments: /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/bin/swift -frontend -c /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/benchmark/single-source/Sim2DArray.swift -target armv7k-apple-watchos2.0 -enable-objc-interop -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/WatchOS.platform/Developer/SDKs/WatchOS3.2.sdk -I /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/swift/benchmark/utils/ObjectiveCTests -I /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/benchmark/O-armv7k-apple-watchos2.0 -F /Applications/Xcode-beta.app/Contents/Developer/Platforms/WatchOS.platform/Developer/SDKs/WatchOS3.2.sdk/../../../Developer/Library/Frameworks -D INTERNAL_CHECKS_ENABLED -emit-module-doc-path /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/benchmark/O-armv7k-apple-watchos2.0/Sim2DArray.swiftdoc -O -parse-as-library -module-name Sim2DArray -emit-module-path /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/benchmark/O-armv7k-apple-watchos2.0/Sim2DArray.swiftmodule -o /Users/buildnode/jenkins/workspace/oss-swift-incremental-RA-osx/buildbot_incremental/swift-macosx-x86_64/benchmark/O-armv7k-apple-watchos2.0/Sim2DArray.o 
1.	While running pass #10319 SILFunctionTransform "SIL Mem2Reg" on SILFunction "@_TTSg5Si___TFSa18_getElementAddressfSiGSpx_".
/bin/sh: line 1: 29000 Abort trap: 6
```

shahmishal commented:
https://ci.swift.org/job/oss-swift-incremental-RA-osx/8423/

eeckstein commented:
Fixed with https://github.com/apple/swift/pull/7558
